### PR TITLE
prov/verbs: Refactor fi_send, fi_rma and atomic functions, add fi_inject functions

### PR DESCRIPTION
**Pull request by a-ilango**
This patch refactors fi_send, fi_rma and atomic functions to use more common code. Added the following fi_inject functions: fi_inject, fi_injectdata, fi_inject_write, fi_inject_writedata. fi_inject_atomic is not supported since verbs doesn't support inline atomic operations.

### CODE REVIEW

File: prov/verbs/src/fi_verbs.c [hunk 1]
```diff
@@ -122,6 +122,13 @@ static struct fi_provider fi_ibv_prov = {
 #define VERBS_INFO_ERRNO(subsys, fn, errno) VERBS_INFO(subsys, fn ": %s(%d)\n",	\
 		strerror(errno), errno)
 
+#define set_sge(sge, buf, len, desc)			\
```
Comment (shefty) [inline prov/verbs/src/fi_verbs.c]: please add a prefix to this to avoid potential conflicts

**Comment by a-ilango** [comment]
We can pass the flags directly but that just moves the code to choose the send_flags from fi_ibv_send to somewhere up the call chain. e.g. If we pass the flags from each of the fi_send calls, then that repeats a lot of code.

File: prov/verbs/src/fi_verbs.c (hunk 1 above)
Comment (a-ilango) [inline prov/verbs/src/fi_verbs.c]: Thanks, will do.

**Comment by shefty** [comment]
But the code isn't passing in FI flags (i.e. FI_INJECT).  It's converting from that to a boolean value that is then converted to the ibv flags. I'm not looking at the integrated code, but conceptually, it seems that there's an unnecessary translation occurring.  For the fi_inject() calls, these two flags could be hard-coded.

**Comment by a-ilango** [comment]
The code isn't passing FI flags because even if the user doesn't pass FI_INJECT, we choose to send messages inline if the size of the message is within the supported inject size. If I understand your comment correctly, the calling function would do something like this:
```
flags |= VERBS_INJECT(ep, len) ? IBV_SEND_INLINE : 0;
flags |= VERBS_COMP(ep) ? IBV_SEND_SIGNALED : 0;
return fi_ibv_send_buf(ep, &wr, buf, len, desc, context, flags);
```
This avoids the translation to an intermediate boolean. The only "drawback" with this approach is the above two lines would be repeated in all send functions instead of just passing the boolean values and setting the flags in fi_ibv_send like in this patch.

**Comment by shefty** [comment]
> flags |= VERBS_INJECT(ep, len) ? IBV_SEND_INLINE : 0;
> flags |= VERBS_COMP(ep) ? IBV_SEND_SIGNALED : 0;
> return fi_ibv_send_buf(ep, &wr, buf, len, desc, context, flags);

basically

> This avoids the translation to an intermediate boolean. The only
> "drawback" with this approach is the above two lines would be repeated in
> all send functions instead of just passing the boolean values and setting
> the flags in fi_ibv_send like in this patch.

But where did the Boolean values come from?  Anywhere there is a hard-coded value of inject = true, you replace with a hard-coded value IBV_SEND_INLINE.  A hard-coded value of inject = false, you replace with 0.  Anywhere else, there must be something like the check above anyway.

**Comment by a-ilango** [comment]
> Anywhere there is a hard-coded value of inject = true, you replace with a hard-coded value IBV_SEND_INLINE.

But that will just repeat the code at many places. Is that OK? Examples would be:
```
return fi_ibv_send_buf(ep, &wr, buf, len, desc, context,
		(VERBS_INJECT(ep, len) ? IBV_SEND_INLINE : 0) | 
                (VERBS_COMP(ep) ? IBV_SEND_SIGNALED : 0))
```
```
return fi_ibv_send_msg(ep, &wr, msg, ((flags & FI_INJECT) ? IBV_SEND_INLINE : 0 ) |
			(VERBS_COMP_FLAGS(ep, flags) ? IBV_SEND_SIGNALED) : 0);
```
```
static ssize_t fi_ibv_send_iov(...) {
...
	if (!(flags & IBV_SEND_SIGNALED))
		flags = VERBS_INJECT(ep, len) ? IBV_SEND_INLINE : 0;
    return fi_ibv_send(ep, wr, len, count, context, flags);
}
```
 May be the different usages above can be wrapped in macros to avoid repetition.

**Comment by shefty** [comment]
> return fi_ibv_send_buf(ep, &wr, buf, len, desc, context,
>         (VERBS_INJECT(ep, len) ? IBV_SEND_INLINE : 0) |
>                 (VERBS_COMP(ep) ? IBV_SEND_SIGNALED : 0))

How does passing in a Boolean value avoid this?  I agree that checking the len in a single place makes sense, which is why I don't see a need for the low level call to take an 'inject' parameter.

**Comment by a-ilango** [comment]
Added a patch to this PR which removes boolean intermediates. Reduced the branches in inject and rma read calls.
@shefty

**Comment by shefty** [comment]
Looks fine to me.

**Pull request merged**